File: 763.划分字母区间.cpp
class Solution {
public:
    vector<int> partitionLabels(string S) {
        int len = S.length();
        vector<int> res;
        if(len == 0)
            return res;
        map<char, int> m;
        int i;
        for(i = 0; i < len; ++i)
            m[S[i]] = i;
        i = 0;
        while(i < len){
            int index = m[S[i]];
            for(int j = i + 1; j < index && j < len; ++j){
                if(m[S[j]] > index)
                    index = m[S[j]];
            }
            res.push_back(index - i + 1);
            i = index + 1;
        }
        return res;
    }
};

File: 763.划分字母区间.java
class Solution {
    public List<Integer> partitionLabels(String S) {
        int[] last = new int[26];
        for(int i = 0; i < S.length(); ++i)
            last[S.charAt(i) - 'a'] = i;
        List<Integer> list = new ArrayList<>();
        int j = 0, index = 0;
        for(int i = 0; i < S.length(); ++i){
            j = Math.max(j, last[S.charAt(i) - 'a']);
            if(i == j){
                list.add(i - index + 1);
                index = i + 1;
            }
        }
        return list;
    }
}